File: Online_Bullying_System_Backend/app/utils/passwords.py
import random
import re
import secrets
import string
from typing import Optional

DEFAULT_SPECIAL_CHARACTERS = "!@#$%^&*()-_=+[]{}<>?/|~"
PASSWORD_SPECIAL_CHARACTERS = "!@#$%^&*()_+-={}[]:;\"'<>.,?/"
COMMON_PASSWORD_PATTERNS = {
    "password",
    "passw0rd",
    "letmein",
    "welcome",
    "admin",
    "root",
    "123456",
    "1234567",
    "12345678",
    "123456789",
    "1234567890",
    "qwerty",
    "abc123",
    "iloveyou",
}
KEYBOARD_SEQUENCES = (
    "qwertyuiop",
    "asdfghjkl",
    "zxcvbnm",
)


def _contains_ascending_sequence(value: str, length: int = 4) -> bool:
    normalized = value.lower()
    for seq in KEYBOARD_SEQUENCES + ("abcdefghijklmnopqrstuvwxyz", "0123456789"):
        for index in range(len(seq) - length + 1):
            if seq[index : index + length] in normalized:
                return True

    for index in range(len(normalized) - length + 1):
        window = normalized[index : index + length]
        if all(ord(window[i + 1]) - ord(window[i]) == 1 for i in range(length - 1)):
            return True
        if all(ord(window[i]) - ord(window[i + 1]) == 1 for i in range(length - 1)):
            return True
    return False


def _contains_repeated_characters(value: str, length: int = 4) -> bool:
    for index in range(len(value) - length + 1):
        window = value[index : index + length]
        if window == window[0] * length:
            return True
    return False


def _contains_personal_information(password: str, user) -> bool:
    if not user:
        return False

    def normalize(value: str) -> str:
        return re.sub(r"[^a-z0-9]", "", value.lower())

    password_normalized = normalize(password)
    raw_values = [
        getattr(user, "full_name", "") or "",
        getattr(user, "email", "") or "",
        getattr(user, "username", "") or "",
    ]
    personal_values = set()
    for raw in raw_values:
        normalized_full = normalize(raw)
        if normalized_full:
            personal_values.add(normalized_full)
        for fragment in re.split(r"[\s@._-]+", str(raw)):
            normalized_fragment = normalize(fragment)
            if len(normalized_fragment) >= 3:
                personal_values.add(normalized_fragment)
    for value in personal_values:
        if value and value in password_normalized:
            return True
    return False


def validate_password_strength(password: str, *, user=None) -> Optional[str]:
    if not password:
        return "Password is required."
    if len(password) < 8:
        return "Password must be at least 8 characters long."
    if not re.search(r"[A-Z]", password):
        return "Password must include at least one uppercase letter."
    if not re.search(r"[a-z]", password):
        return "Password must include at least one lowercase letter."
    if not re.search(r"[0-9]", password):
        return "Password must include at least one number."
    special_characters_pattern = re.compile(f"[{re.escape(PASSWORD_SPECIAL_CHARACTERS)}]")
    if not special_characters_pattern.search(password):
        return (
            "Password must include at least one special character "
            "(! @ # $ % ^ & * ( ) _ + - = { } [ ] : ; \" ' < > , . ? /)."
        )

    lowered = password.lower()
    if lowered in COMMON_PASSWORD_PATTERNS:
        return "Password is too common. Choose something harder to guess."
    for pattern in COMMON_PASSWORD_PATTERNS:
        if pattern in lowered:
            return "Password should not contain common words like 'password' or '123456'."

    if _contains_personal_information(password, user):
        return "Password must not contain your personal information."

    if _contains_ascending_sequence(password, 4):
        return "Password must not contain sequential patterns like 'abcd' or '1234'."

    if _contains_repeated_characters(password, 4):
        return "Password must not contain repeated characters like '1111'."

    return None


def generate_strong_password(length: int = 12, special_chars: str = DEFAULT_SPECIAL_CHARACTERS) -> str:
    """
    Generate a cryptographically secure password that satisfies the following:
      * Contains at least one uppercase letter
      * Contains at least one lowercase letter
      * Contains at least one digit
      * Contains at least one special character
    """
    length = max(length, 8)
    if not special_chars:
        special_chars = DEFAULT_SPECIAL_CHARACTERS

    categories = [
        secrets.choice(string.ascii_uppercase),
        secrets.choice(string.ascii_lowercase),
        secrets.choice(string.digits),
        secrets.choice(special_chars),
    ]

    remaining_length = length - len(categories)
    if remaining_length > 0:
        all_characters = string.ascii_letters + string.digits + special_chars
        categories.extend(secrets.choice(all_characters) for _ in range(remaining_length))

    sys_random = random.SystemRandom()
    sys_random.shuffle(categories)
    return "".join(categories)

File: Online_Bullying_System_Frontend/src/utils/passwords.js
const PASSWORD_SPECIAL_CHARACTERS = '!@#$%^&*()_+-={}[]:;"\'<>.,?/';
const COMMON_PASSWORD_PATTERNS = [
  'password',
  'passw0rd',
  'letmein',
  'welcome',
  'admin',
  'root',
  '123456',
  '1234567',
  '12345678',
  '123456789',
  '1234567890',
  'qwerty',
  'abc123',
  'iloveyou',
];
const KEYBOARD_SEQUENCES = ['qwertyuiop', 'asdfghjkl', 'zxcvbnm'];

const escapeRegExp = (value) => value.replace(/[.*+?^${}()|[\]\\-]/g, '\\$&');
const SPECIAL_CHAR_REGEX = new RegExp(`[${escapeRegExp(PASSWORD_SPECIAL_CHARACTERS)}]`);

const normalizePersonalValue = (value) =>
  (value || '').toLowerCase().replace(/[^a-z0-9]/g, '');

const containsPersonalInfo = (password, context = {}) => {
  const passwordNormalized = normalizePersonalValue(password);
  if (!passwordNormalized) {
    return false;
  }
  const rawValues = [context.fullName, context.email, context.username].filter(Boolean);
  const personalValues = [];
  rawValues.forEach((raw) => {
    const normalizedFull = normalizePersonalValue(raw);
    if (normalizedFull) {
      personalValues.push(normalizedFull);
    }
    String(raw)
      .split(/[\s@._-]+/)
      .forEach((fragment) => {
        const normalizedFragment = normalizePersonalValue(fragment);
        if (normalizedFragment && normalizedFragment.length >= 3) {
          personalValues.push(normalizedFragment);
        }
      });
  });
  return personalValues.some((entry) => passwordNormalized.includes(entry));
};

const hasSequentialCharacters = (value, length = 4) => {
  if (!value) return false;
  const normalized = value.toLowerCase();
  const sequences = [...KEYBOARD_SEQUENCES, 'abcdefghijklmnopqrstuvwxyz', '0123456789'];

  for (const sequence of sequences) {
    for (let index = 0; index <= sequence.length - length; index += 1) {
      if (normalized.includes(sequence.slice(index, index + length))) {
        return true;
      }
    }
  }

  for (let index = 0; index <= normalized.length - length; index += 1) {
    const window = normalized.slice(index, index + length);
    let ascending = true;
    let descending = true;
    for (let offset = 0; offset < window.length - 1; offset += 1) {
      if (window.charCodeAt(offset + 1) - window.charCodeAt(offset) !== 1) {
        ascending = false;
      }
      if (window.charCodeAt(offset) - window.charCodeAt(offset + 1) !== 1) {
        descending = false;
      }
    }
    if (ascending || descending) {
      return true;
    }
  }
  return false;
};

const hasRepeatedCharacters = (value, length = 4) => {
  if (!value) return false;
  for (let index = 0; index <= value.length - length; index += 1) {
    const window = value.slice(index, index + length);
    if (window === window[0].repeat(length)) {
      return true;
    }
  }
  return false;
};

const validateNewPassword = (password, context = {}) => {
  if (!password) {
    return 'Password is required.';
  }
  if (password.length < 8) {
    return 'Password must be at least 8 characters long.';
  }
  if (!/[A-Z]/.test(password)) {
    return 'Password must include at least one uppercase letter.';
  }
  if (!/[a-z]/.test(password)) {
    return 'Password must include at least one lowercase letter.';
  }
  if (!/[0-9]/.test(password)) {
    return 'Password must include at least one number.';
  }
  if (!SPECIAL_CHAR_REGEX.test(password)) {
    return 'Password must include at least one special character (! @ # $ % ^ & * ( ) _ + - = { } [ ] : ; " \' < > , . ? /).';
  }
  const lowered = password.toLowerCase();
  if (COMMON_PASSWORD_PATTERNS.includes(lowered)) {
    return 'Password is too common. Choose something harder to guess.';
  }
  if (COMMON_PASSWORD_PATTERNS.some((pattern) => lowered.includes(pattern))) {
    return "Password should not contain common words like 'password' or '123456'.";
  }
  if (containsPersonalInfo(password, context)) {
    return 'Password must not contain your personal information.';
  }
  if (hasSequentialCharacters(password)) {
    return "Password must not contain sequential patterns like 'abcd' or '1234'.";
  }
  if (hasRepeatedCharacters(password)) {
    return "Password must not contain repeated characters like '1111'.";
  }
  return null;
};

const evaluatePasswordRules = (password, context = {}) => {
  const value = typeof password === 'string' ? password : '';
  const hasValue = value.length > 0;
  const meetsLength = value.length >= 8;
  const hasUppercase = /[A-Z]/.test(value);
  const hasLowercase = /[a-z]/.test(value);
  const hasDigit = /[0-9]/.test(value);
  const hasSpecial = SPECIAL_CHAR_REGEX.test(value);
  const lowered = value.toLowerCase();
  const containsCommonPattern =
    hasValue && COMMON_PASSWORD_PATTERNS.some((pattern) => lowered.includes(pattern));
  const containsPersonalDetails = hasValue && containsPersonalInfo(value, context);
  const containsSequential = hasValue && hasSequentialCharacters(value);
  const containsRepeated = hasValue && hasRepeatedCharacters(value);

  return {
    length: meetsLength,
    uppercase: hasUppercase,
    lowercase: hasLowercase,
    digit: hasDigit,
    special: hasSpecial,
    common: hasValue && !containsCommonPattern && !containsPersonalDetails,
    sequence: hasValue && !containsSequential && !containsRepeated,
  };
};

const PASSWORD_RULE_ITEMS = [
  { id: 'length', label: 'At least 8 characters.' },
  { id: 'uppercase', label: 'At least one uppercase letter (A–Z).' },
  { id: 'lowercase', label: 'At least one lowercase letter (a–z).' },
  { id: 'digit', label: 'At least one number (0–9).' },
  { id: 'special', label: 'At least one special character' },
];

const getPasswordRuleChecklist = (password, context = {}) => {
  const status = evaluatePasswordRules(password, context);
  return PASSWORD_RULE_ITEMS.map((rule) => ({
    ...rule,
    met: Boolean(status[rule.id]),
  }));
};

export {
  PASSWORD_SPECIAL_CHARACTERS,
  COMMON_PASSWORD_PATTERNS,
  SPECIAL_CHAR_REGEX,
  validateNewPassword,
  evaluatePasswordRules,
  getPasswordRuleChecklist,
};
